crates/sodium/src/crypto_system/secure_buffer.rs

```diff
@@ -0,0 +1,130 @@
+use lib3h_crypto_api::{Buffer, BufferType, CryptoError, CryptoResult, ProtectState};
+
+use crate::check_init;
+use libc::c_void;
+
+/// A secure buffer implementation of lib3h_crypto_api::Buffer
+/// making use of libsodium's implementation of mlock and mprotect.
+pub struct SecureBuffer {
+    z: *mut c_void,
+    s: usize,
+    p: std::cell::RefCell<ProtectState>,
+}
+
+unsafe impl Send for SecureBuffer {}
+
+impl Drop for SecureBuffer {
+    fn drop(&mut self) {
+        unsafe {
+            rust_sodium_sys::sodium_free(self.z);
+        }
+    }
+}
+
+impl Clone for SecureBuffer {
+    fn clone(&self) -> Self {
+        let mut out = SecureBuffer::new(self.s).expect("could not alloc new");
+        out.write(0, &self.read_lock())
+            .expect("could not write new");
+        out
+    }
+}
+
+impl std::fmt::Debug for SecureBuffer {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self.p.borrow() {
+            ProtectState::NoAccess => write!(f, "SecureBuffer( {:?} )", "<NO_ACCESS>"),
+            _ => write!(f, "SecureBuffer( {:?} )", *self),
+        }
+    }
+}
+
+impl std::ops::Deref for SecureBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        if *self.p.borrow() == ProtectState::NoAccess {
+            panic!("Deref, but state is NoAccess");
+        }
+        unsafe { &std::slice::from_raw_parts(self.z as *const u8, self.s)[..self.s] }
+    }
+}
+
+impl std::ops::DerefMut for SecureBuffer {
+    fn deref_mut(&mut self) -> &mut [u8] {
+        if *self.p.borrow() != ProtectState::ReadWrite {
+            panic!("DerefMut, but state is not ReadWrite");
+        }
+        unsafe { &mut std::slice::from_raw_parts_mut(self.z as *mut u8, self.s)[..self.s] }
+    }
+}
+
+impl BufferType for SecureBuffer {}
+
+impl Buffer for SecureBuffer {
+    fn new(size: usize) -> CryptoResult<Self> {
+        check_init();
+        let z = unsafe {
+            // round up to nearest 8
+            let align_size = (size + 7) & !7;
```

Comment:
What's this magic doing? ^^'

Reply:
sodium_malloc has some undefined behavior if you give it a size that is not memory-aligned (https://download.libsodium.org/doc/memory_management#guarded-heap-allocations), so we need to round up to the nearest 8 bytes.